src/plugins/crypto/gpg.c

```diff
@@ -0,0 +1,439 @@
+		{
+			recipientCount++;
+		}
+	}
+
+	if (recipientCount == 0)
+	{
+		ELEKTRA_SET_ERRORF (ELEKTRA_ERROR_CRYPTO_CONFIG_FAULT, errorKey,
+				    "Missing GPG key (specified as %s) in plugin configuration.", ELEKTRA_CRYPTO_PARAM_GPG_KEY);
+		return -1;
+	}
+
+	if (inTestMode (conf))
+	{
+		// add two parameters for unit testing
+		testMode = 2;
+	}
+
+	// initialize argument vector for gpg call
+	const kdb_unsigned_short_t argc = (2 * recipientCount) + 3 + testMode;
+	kdb_unsigned_short_t i = 1;
+	char * argv[argc];
+
+	// append root (crypto/key) as gpg recipient
+	if (root && strlen (keyString (root)) > 0)
+	{
+		argv[i] = "-r";
+		// NOTE argv[] values will not be modified, so const can be discarded safely
+		argv[i + 1] = (char *)keyString (root);
+		i = i + 2;
+	}
+
+	// append keys beneath root (crypto/key/#_) as gpg recipients
+	ksRewind (conf);
+	while ((k = ksNext (conf)) != 0)
+	{
+		if (keyIsBelow (k, root))
+		{
+			argv[i] = "-r";
+			// NOTE argv[] values will not be modified, so const can be discarded safely
+			argv[i + 1] = (char *)keyString (k);
+			i = i + 2;
+		}
+	}
+
+	// append option for unit tests
+	if (testMode)
+	{
+		argv[argc - 4] = "--trust-model";
+		argv[argc - 3] = "always";
+	}
+
+	argv[argc - 2] = "-e";
+
+	// call gpg
+	return elektraCryptoGpgCall (conf, errorKey, msgKey, argv, argc);
+}
+
+/**
+ * @brief call the gpg binary to decrypt the random master password r.
+ *
+ * @param conf holds the backend/plugin configuration
+ * @param errorKey holds the error description in case of failure
+ * @param msgKey holds the master password to be decrypted. Note that the content of this key will be modified.
+ *
+ * @retval 1 on success
+ * @retval -1 on failure
+ */
+int elektraCryptoGpgDecryptMasterPassword (KeySet * conf, Key * errorKey, Key * msgKey)
+{
+	if (inTestMode (conf))
+	{
+		char * argv[] = { "", "--trust-model", "always", "-d", NULL };
+		return elektraCryptoGpgCall (conf, errorKey, msgKey, argv, 5);
+	}
+	else
+	{
+		char * argv[] = { "", "-d", NULL };
+		return elektraCryptoGpgCall (conf, errorKey, msgKey, argv, 3);
+	}
+}
+
+/**
+ * @brief call the gpg binary to perform the requested operation.
+ *
+ * @param conf holds the backend/plugin configuration
+ * @param errorKey holds the error description in case of failure
+ * @param msgKey holds the message to be transformed
+ * @param argv array holds the arguments passed on to the gpg process
+ * @param argc contains the number of elements in argv, i.e. the size of argv
+ *
+ * @retval 1 on success
+ * @retval -1 on failure
+ */
+int elektraCryptoGpgCall (KeySet * conf, Key * errorKey, Key * msgKey, char * argv[], size_t argc)
+{
+	pid_t pid;
+	int status;
+	int pipe_stdin[2];
+	int pipe_stdout[2];
+	kdb_octet_t * buffer = NULL;
+	const ssize_t bufferSize = 2 * keyGetValueSize (msgKey);
+	ssize_t outputLen;
+
+	assert (argc > 1);
+
+	// sanitize the argument vector
+	if (getGpgBinary (&argv[0], conf, errorKey) != 1)
+	{
+		return -1;
```

Comment:
No error set here, but it seems like (at least some) callers expect the error already to be set. It would be good if you can document who is responsible for setting errors.